sdk/cosmosdb/cosmos/src/bulk/BulkStreamer.ts

```diff
@@ -0,0 +1,267 @@
+import { Constants, getPathFromLink, ResourceType } from "../common";
+import { BulkResponse } from "./BulkResponse";
+import { ItemBulkOperation } from "./ItemBulkOperation";
+import { addDignosticChild } from "../utils/diagnostics";
+import type { BulkOperationResult } from "./BulkOperationResult";
+import { BulkExecutionRetryPolicy } from "../retry/bulkExecutionRetryPolicy";
+import type { RetryPolicy } from "../retry/RetryPolicy";
+import { Limiter } from "./Limiter";
+
+/**
+ * BulkStreamer for bulk operations in a container.
+ * It maintains one @see {@link BulkStreamerPerPartition} for each Partition Key Range, which allows independent execution of requests. Semaphores are in place to rate limit the operations
+ * at the Streamer / Partition Key Range level, this means that we can send parallel and independent requests to different Partition Key Ranges, but for the same Range, requests
+ * will be limited. Two callback implementations define how a particular request should be executed, and how operations should be retried. When the streamer dispatches a batch
+ * the batch will create a request and call the execute callback (executeRequest), if conditions are met, it might call the retry callback (reBatchOperation).
+ * @hidden
+ */
+
+export class BulkStreamer {
+  private readonly container: Container;
+  private readonly clientContext: ClientContext;
+  private readonly partitionKeyRangeCache: PartitionKeyRangeCache;
+  private readonly streamersByPartitionKeyRangeId: Map<string, BulkStreamerPerPartition>;
+  private readonly limitersByPartitionKeyRangeId: Map<string, Limiter>;
+  private options: RequestOptions;
+  private orderedResponse: BulkOperationResult[] = [];
+  private diagnosticNode: DiagnosticNodeInternal;
+  private operationPromises: Promise<BulkOperationResult>[] = [];
+  private operationIndex: number = 0;
+
+  /**
+   * @internal
+   */
+  constructor(
+    container: Container,
+    clientContext: ClientContext,
+    partitionKeyRangeCache: PartitionKeyRangeCache,
+  ) {
+    this.container = container;
+    this.clientContext = clientContext;
+    this.partitionKeyRangeCache = partitionKeyRangeCache;
+    this.streamersByPartitionKeyRangeId = new Map();
+    this.limitersByPartitionKeyRangeId = new Map();
+
+    this.executeRequest = this.executeRequest.bind(this);
+    this.reBatchOperation = this.reBatchOperation.bind(this);
+  }
+  /**
+   *
+   * @internal
+   */
+  initializeBulk(options: RequestOptions): void {
+    this.orderedResponse = [];
+    this.options = options;
+    this.operationIndex = 0;
+    this.operationPromises = [];
+    this.diagnosticNode = new DiagnosticNodeInternal(
+      this.clientContext.diagnosticLevel,
+      DiagnosticNodeType.CLIENT_REQUEST_NODE,
+      null,
+    );
+  }
+
+  /**
+   * adds operation(s) to the streamer
+   * @param operationInput - bulk operation or list of bulk operations
+   */
+  addOperations(operationInput: OperationInput | OperationInput[]): void {
```

Comment:
we should allow only OperationInput[] in the argument.

Reply:
Any reasoning behind this? we discussed this to allow supporting both inputs.